[CodeComplete] Add test case for completing in `inout` arguments

ahoppen opened the pull request:
This captures a crash found in rdar://74786497 that no longer occurs in `main`.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Windows

ahoppen commented:
@swift-ci Please test Windows

xedin commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -1104,6 +1104,10 @@ fallbackTypeCheck(DeclContext *DC) {
 }
 
 static Type getTypeForCompletion(const constraints::Solution &S, Expr *E) {
+  if (!S.hasType(E)) {
```
If the expression was passed to `typeCheckForCodeCompletion` it should have a type set always, I wonder if this is bug in code completion or constraint solver?

hamishknight replied:
@xedin Unfortunately it seems we're type checking the wrong expression somewhere which causes us to fail to get the type of the original code completion expr, this check at least stops us from crashing in that case.

In addition for examples like completion in a result builder, which I believe Alex is currently working on, it's possible the completion expr is in an invalid statement that gets marked an unhandled and therefore never reaches CSGen. It would be nice if we could better recover from that with holes, but for now this lets us recover.

xedin replied:
It sounds like the actual problem is that there is no feedback to distinguish invalid cases. It makes me think that the actual fix here might be to register invalid expressions/statements as having `ErrorType` in the constraint system.

And sorry, for some reason github didn't notify me that there was a rely here...

ahoppen committed:
[CodeComplete] Add test case for completing in `inout` arguments

This captures a crash found in rdar://74786497 that no longer occurs in `main`.

ahoppen commented:
I found a reproducer for the crasher. It no longer occurs in `main`, so I’m re-purposing this PR to just add a test case for it.

ahoppen commented:
@swift-ci Please test

xedin commented:
Thank you, @ahoppen!

ahoppen commented:
@swift-ci Please test Linux
